go-runtime/compile/schema.go

```diff
@@ -201,56 +204,143 @@ func visitCallExpr(pctx *parseContext, node *ast.CallExpr) {
-			var err error
-			name, err = strconv.Unquote(literal.Value)
-			if err != nil {
-				pctx.errors.add(wrapf(node, err, ""))
-				return
-			}
+func parseFSMDecl(pctx *parseContext, node *ast.CallExpr) {
+	var literal *ast.BasicLit
+	if len(node.Args) > 0 {
+		literal, _ = node.Args[0].(*ast.BasicLit)
+	}
+	if literal == nil || literal.Kind != token.STRING {
+		pctx.errors.add(errorf(node, "first argument to an FSM declaration must be the name as a string literal"))
+		return
+	}
+
+	name, err := strconv.Unquote(literal.Value)
+	if err != nil {
+		panic(err) // Should never happen
+	}
+
+	if !schema.ValidateName(name) {
+		pctx.errors.add(errorf(node, "FSM names must be valid identifiers"))
+	}
+
+	fsm := &schema.FSM{
+		Pos:  goPosToSchemaPos(node.Pos()),
+		Name: name,
+	}
+	pctx.module.Decls = append(pctx.module.Decls, fsm)
+
+	for _, arg := range node.Args[1:] {
+		call, ok := arg.(*ast.CallExpr)
+		if !ok {
+			pctx.errors.add(errorf(arg, "expected call to Start or Transition"))
+			continue
 		}
+		_, fn := deref[*types.Func](pctx.pkg, call.Fun)
+		if fn == nil {
+			pctx.errors.add(errorf(call, "expected call to Start or Transition"))
+			continue
+		}
+		parseFSMTransition(pctx, call, fn, fsm)
 	}
-	if name == "" {
-		pctx.errors.add(errorf(node, "config and secret declarations must have a single string literal argument"))
+}
+
+// Parse a Start or Transition call in an FSM declaration and add it to the FSM.
+func parseFSMTransition(pctx *parseContext, node *ast.CallExpr, fn *types.Func, fsm *schema.FSM) {
+	refs := make([]*schema.Ref, len(node.Args))
+	for i, arg := range node.Args {
+		ref := parseVerbRef(pctx, arg)
+		if ref == nil {
+			pctx.errors.add(errorf(arg, "expected a reference to a sink"))
+			return
+		}
+		refs[i] = ref
+	}
+	switch fn.FullName() {
+	case ftlStartFuncPath:
+		if len(refs) != 1 {
+			pctx.errors.add(errorf(node, "expected one reference to a sink"))
+			return
+		}
+		fsm.Start = append(fsm.Start, refs...)
+
+	case ftlTransitionFuncPath:
+		if len(refs) != 2 {
+			pctx.errors.add(errorf(node, "expected two references to sinks"))
+			return
+		}
+		fsm.Transitions = append(fsm.Transitions, &schema.FSMTransition{
+			Pos:  goPosToSchemaPos(node.Pos()),
+			From: refs[0],
+			To:   refs[1],
+		})
+
+	default:
+		pctx.errors.add(errorf(node, "expected call to Start or Transition"))
+	}
+}
+
+func parseConfigDecl(pctx *parseContext, node *ast.CallExpr, fn *types.Func) {
+	var literal *ast.BasicLit
+	if len(node.Args) > 0 {
+		literal, _ = node.Args[0].(*ast.BasicLit)
+	}
+	if literal == nil || literal.Kind != token.STRING {
+		pctx.errors.add(errorf(node, "first argument to config and secret declarations must be the name as a string literal"))
 		return
 	}
+	name, err := strconv.Unquote(literal.Value)
+	if err != nil {
+		panic(err) // Should never happen
+	}
+
+	if !schema.ValidateName(name) {
+		pctx.errors.add(errorf(node, "config and secret names must be valid identifiers"))
```

Comment:
Could we incorporate some context on what constitutes a valid identifier (i.e. alphanumeric string)? I've often gotten a mild case of the grumps from error messages that say something is invalid, but not what makes it invalid. Maybe we could switch the `bool` return type to an `error` (e.g. `fmt.Errorf("%q is not an alphanumeric string", name)`) and use `err.Error()` here?